# NLP - HW6
### Miguel Bonilla

- [1. Evaluate Text Similarity](#1.-Evaluate-Text-Similarity)
    - [a. Do a Book Search on Amazon](#a.-Do-a-Book-Search-on-Amazon)
    - [b. Run a Text Similarity Measure](#b.-Run-a-Text-Similarity-Measure)
    - [c. Which Two Texts Are the Most Similar?](#c.-Which-Two-Texts-Are-the-Most-Similar?)
- [2. Evaluate Using Major Search Engine](#2.-Evaluate-Using-Major-Search-Engine)
    - [a. Enter a Book Title from 1a into Google](#a.-Enter-a-Book-Title-from-1a-into-Google)
    - [b. Run the Same Text Similarity from 1b](#b.-Run-the-Same-Text-Similarity-from-1b)
    - [c. Which One Has the Highest Similarity?](#c.-Which-One-Has-the-Highest-Similarity?)
- [Summary](#Summary)

1. Evaluate text similarity of Amazon book search results by doing the following:  
a. Do a book search on Amazon via the search box. Manually copy the full book title 
(including subtitle) of each of the top 24 books listed in the first two pages of search 
results. You need to share the search query you use.  
b. In Python, run one of the text-similarity measures covered in this course, e.g., cosine 
similarity. Compare each of the book titles, pairwise, to every other one.  
c. Which two titles are the most similar to each other? Which are the most dissimilar? 
Where do they rank, among the first 24 results?  
2. Now evaluate using a major search engine.  
a. Enter one of the book titles from question 1a into Google, Bing, or Yahoo!. Copy the 
capsule of the first organic result and the 20th organic result. Take web results only (i.e.,
not video results), and skip sponsored results.  
b. Run the same text similarity calculation that you used for question 1b on each of these 
capsules in comparison to the original query (book title).  
c. Which one has the highest similarity measure?  

Submit all of your inputs and outputs and your code for this assignment, along with a brief written 
explanation of your findings. 

In [1]:
import pandas as pd
import nltk
import numpy as np
from nltk.tokenize import word_tokenize
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

### 1. Evaluate Text Similarity

Titles saved manually from an Amazon.com books search for "Chronicle of a Death Foretold". The first 24 non-sponsored results were saved into a csv file and then loaded into a pandas dataframe.

#### a. Do a Book Search on Amazon

In [2]:
titles = pd.read_csv('https://raw.githubusercontent.com/boneeyah/DS7337/main/Book_titles.csv')
titles = titles['Book title'].to_list()
titles

['Chronicle of a Death Foretold',
 'Chronicle of a Death Foretold - Teacher Guide by Novel Units',
 'Chronicle of a Death Foretold 1st (first) edition Text Only',
 'Chronicle of a Death Foretold (Chinese Edition)',
 'Chronicle of a Death Foretold (SparkNotes Literature Guide) (SparkNotes Literature Guide Series)',
 'Crónica de una muerte anunciada / Chronicle of a Death Foretold (Spanish Edition)',
 'Chronicle of a Death Foretold by Gabriel García Márquez (Book Analysis): Detailed Summary, Analysis and Reading Guide (BrightSummaries.com)',
 'Crónica de una muerte anunciada [Chronicle of a Death Foretold]',
 'GradeSaver(tm) ClassicNotes Chronicle of a Death Foretold',
 'Destiny & A Chronicle of Deaths Foretold (Books One, Two and Three)',
 'Cien años de soledad (50 Aniversario) / One Hundred Years of Solitude: Illustrated Fiftieth Anniversary edition of One Hundred Years of Solitude (Spanish Edition)',
 'A Study Guide for Gabriel Garcia Marquez\'s "Chronicle of a Death Foretold" (For St

#### b. Run a Text-Similarity Measure

In [3]:
### special characters to be removed
### We won't be removing stop words, since titles are made up of content words
#stop_words = nltk.corpus.stopwords.words('english') + nltk.corpus.stopwords.words('spanish')
special = [':',',','-','(',')','[',']','–','/','#','``',';','.','&','"',"''"]

In [4]:
def normalize_list(title_list):
    term = [word_tokenize(term.lower()) for term in title_list]
    blank_list = []
    for i in range(len(title_list)):
        blank_list.append(' '.join([w for w in term[i] if w not in special]))
    return(np.array(blank_list))

In [5]:
norm_corpus = normalize_list(titles)

In [6]:
norm_corpus

array(['chronicle of a death foretold',
       'chronicle of a death foretold teacher guide by novel units',
       'chronicle of a death foretold 1st first edition text only',
       'chronicle of a death foretold chinese edition',
       'chronicle of a death foretold sparknotes literature guide sparknotes literature guide series',
       'crónica de una muerte anunciada chronicle of a death foretold spanish edition',
       'chronicle of a death foretold by gabriel garcía márquez book analysis detailed summary analysis and reading guide brightsummaries.com',
       'crónica de una muerte anunciada chronicle of a death foretold',
       'gradesaver tm classicnotes chronicle of a death foretold',
       'destiny a chronicle of deaths foretold books one two and three',
       'cien años de soledad 50 aniversario one hundred years of solitude illustrated fiftieth anniversary edition of one hundred years of solitude spanish edition',
       "a study guide for gabriel garcia marquez 's ch

In [7]:
tf = TfidfVectorizer(ngram_range=(1, 2), min_df=0, analyzer='char_wb')
tfidf_matrix = tf.fit_transform(norm_corpus)

In [8]:
index = []
for i in range(len(titles)):
    index.append('Book{}'.format(i+1))

In [9]:
doc_sim = cosine_similarity(tfidf_matrix)
doc_sim_df = pd.DataFrame(doc_sim,index=index,columns=index)
doc_sim_df.head()

Book1     Book2     Book3     Book4     Book5     Book6     Book7  \
Book1  1.000000  0.858996  0.809811  0.898119  0.697976  0.831611  0.686730   
Book2  0.858996  1.000000  0.752948  0.811050  0.748553  0.808989  0.738397   
Book3  0.809811  0.752948  1.000000  0.810859  0.629626  0.744778  0.621437   
Book4  0.898119  0.811050  0.810859  1.000000  0.696957  0.833387  0.659242   
Book5  0.697976  0.748553  0.629626  0.696957  1.000000  0.718408  0.652015   

          Book8     Book9    Book10  ...    Book15    Book16    Book17  \
Book1  0.842867  0.822993  0.865894  ...  0.697705  0.793894  0.816838   
Book2  0.812901  0.798692  0.820105  ...  0.667756  0.779060  0.749092   
Book3  0.704483  0.706205  0.774129  ...  0.616255  0.750159  0.758016   
Book4  0.782577  0.773853  0.832231  ...  0.681862  0.784950  0.775979   
Book5  0.677527  0.726433  0.683920  ...  0.592820  0.680363  0.623737   

         Book18    Book19    Book20    Book21    Book22    Book23    Book24  
Book1  0.824996  0.946549  0.824996  0.874578  0.858359  0.870948  0.822371  
Book2  0.756573  0.847260  0.756573  0.863140  0.842993  0.775280  0.813160  
Book3  0.748221  0.833449  0.748221  0.773665  0.759215  0.741715  0.726851  
Book4  0.783729  0.856036  0.783729  0.831679  0.820889  0.816846  0.794915  
Book5  0.629966  0.696257  0.629966  0.709271  0.695291  0.683941  0.712227  

[5 rows x 24 columns]

In [10]:
doc_sim_df.to_csv('matrix_q1.csv')

#### c. Which Two Texts Are the Most Similar?

In [11]:
x=[]
for i in range(24):
    x.append(25*i)

arr = np.argsort(-doc_sim.flatten())
np.setdiff1d(arr,x,assume_unique=True)[0:4]

array([181, 319, 304, 396], dtype=int64)

In [21]:
score1 = doc_sim.flatten()[181]
score1

0.9897322499673622

In [23]:
modulo = 181 % 24
floor_div = 181 // 24
print('The two books that are the most similar are Book {} and Book {}\n which have a similarity score of {}'.format(floor_div+1,modulo+1,score1))

The two books that are the most similar are Book 8 and Book 14
 which have a similarity score of 0.9897322499673622


In [14]:
print('The original book titles prior to normalization are :\n "{}" and "{}"'.format(titles[floor_div],titles[modulo]))

The original book titles prior to normalization are :
 "Crónica de una muerte anunciada [Chronicle of a Death Foretold]" and "Crónica de una muerte Anunciada [Chronicle of a Death Foretold] Cronica"


### 2. Evaluate Using Major Search Engine

#### a. Enter a Book Title from 1a into Google

Searching on Google.com for the title of the first result that was returned on 1 a. - Chronicle of a Death Foretold.

In [15]:
### first result on google
first_res = 'Chronicle of a Death Foretold - Amazon.com'
twentieth_res = 'Chronicle of a Death Foretold Summary & Study Guide'
original_query = 'Chronicle of a Death Foretold'
new_titles = [first_res,twentieth_res, original_query]
new_index = ['First','Twentieth','Original']

#### b. Run the Same Text Similarity from 1b

In [16]:
norm_new = normalize_list(new_titles)
norm_new

array(['chronicle of a death foretold amazon.com',
       'chronicle of a death foretold summary study guide',
       'chronicle of a death foretold'], dtype='<U49')

In [17]:
tfidf_matrix2 = tf.fit_transform(norm_new)
new_sim = cosine_similarity(tfidf_matrix2)
new_sim_df = pd.DataFrame(new_sim, columns=new_index, index = new_index)

In [18]:
new_sim_df

First  Twentieth  Original
First      1.000000   0.809371  0.924832
Twentieth  0.809371   1.000000  0.855151
Original   0.924832   0.855151  1.000000

In [19]:
new_sim_df.to_csv('matrix_q2.csv')

#### c. Which One Has the Highest Similarity?

The first Google search result has a higher similarity score with the book title from the original query (Chronicle of a Death Foretold). The first Google search result had a similarity score of .9248, compared to a score of .8551 for the twentieth result.

### Summary

The overall approach of this project was as follows:  

First, a list of titles for the first 20 results from an Amazon Books search for "Chronicle of a Death Foretold" were manually collected and saved into a csv file.  

Then, these titles were then normalized by removing special characters and lower casing all words, stop words were not removed during normalization since titles will be taken as entities, and as such all words were deemed to be content words. If the titles had been normalized, "Chronicle of a Death Foretold" and "Chronicle of a Death Not Foretold" would have yieleded a perfect match (Not would have been removed), when we know these are in fact entirely different works. Once normalized, the titles were vectorized into a 24x24 matrix, and the cosine similarities calculated for each pairwise comparison using the cosine_similarity function from the scikitlearn package.  

Then, the most similar books were found by removing the diagonal values from the matrix (as the similarity of each book title with itself is 1), and finding the pair with the highest cosine similarity score. The two most similar results were "Crónica de una muerte anunciada [Chronicle of a Death Foretold]" and "Crónica de una muerte Anunciada [Chronicle of a Death Foretold] Cronica" which differed in a single extra word on one of the titles. The cosine similarity score for the two most similar titles was .9897.

Subsequently, the same calculations were calculated using the first and twentieth results from a Google Search for the title "Chronicle of a Death Foretold". The results had a higher similarity score for the first result, "Chronicle of a Death Foretold - Amazon.com"; than the twentieth result, "Chronicle of a Death Foretold Summary & Study Guide". Meaning the first result was not suprisingly a better match to our search than the twentieth result, with cosine similarity scores of .9248 and .8551, respectively.